Remove native_functions.yaml dependency from TensorTopK.cu

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67721
* #66980
* #66979
* #67216
* #66914
* #66913
* __->__ #66794
* #66793

Differential Revision: [D31856104](https://our.internmc.facebook.com/intern/diff/D31856104)

peterbell10 committed:
Remove native_functions.yaml dependency from TensorTopK.cu

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorTopK.cu"




[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorTopK.cu"




[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorTopK.cu"




[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorTopK.cu"


Differential Revision: [D31856104](https://our.internmc.facebook.com/intern/diff/D31856104)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorTopK.cu"


Differential Revision: [D31856104](https://our.internmc.facebook.com/intern/diff/D31856104)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from TensorTopK.cu"


Differential Revision: [D31856104](https://our.internmc.facebook.com/intern/diff/D31856104)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856104).

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorTopK.cu"


Differential Revision: [D31856104](https://our.internmc.facebook.com/intern/diff/D31856104)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorTopK.cu"


Differential Revision: [D31856104](https://our.internmc.facebook.com/intern/diff/D31856104)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856104).

dagitses commented on aten/src/ATen/native/cuda/TensorTopK.cu:
```diff
@@ -161,26 +163,15 @@ __global__ void gatherTopK(at::cuda::detail::TensorInfo<T, IndexType> input,
+void launch_gather_topk_kernel(
+    const TensorBase& self, int64_t k, int64_t dim, bool largest, bool sorted,
+    const TensorBase& values, const TensorBase& indices) {
   int numDims = self.dim();
```
out of curiosity, what is the criteria by which these four lines remain in the .cu but the `if (k == 0)` moves?

My understanding of your work is that you are prioritizing the following:
1) pull out dependencies of Tensor and ATen.h from .cu files.
2) less important but where possible, extract other unnecessary dependencies from .cu files
3) take advantage of other opportunities to move code out of .cu files

Is that roughly the prioritization of your approach here? Under that, moving this block and the `k == 0` check both fit under 3) as the lowest priority.

Reordering the `k == 0` check does change the behavior since it now avoids the check about having too many dimensions. Is that OK? FWIW, I like the idea of being stringent on inputs rather than letting a loophole like this let the user get away with an invalid input.


Changing topic altogether, do you think that splitting code up this way causes any meaningful harm by creating cross-module optimization barriers?

peterbell10 replied:
>out of curiosity, what is the criteria by which these four lines remain in the .cu but the `if (k == 0)` moves?

The `return` statement must go in the `.cpp` file function so we don't launch the sorting kernels. It would make sense to keep the `MAX_DIMS` checks together with it, but `MAX_DIMS` is defined in a `.cuh` header file so needs `nvcc`:
https://github.com/pytorch/pytorch/blob/383c1f51b156fab7d069f011113e2934642ccb93/aten/src/ATen/cuda/detail/OffsetCalculator.cuh#L19

>My understanding of your work is that you are prioritizing the following: [...]

This is mostly right, although `.cu` isn't actually in my criteria anywhere. I'm currently focusing on files that depend on `native_functions.yaml`, prioritized by their compile time (to maximize impact). It just so happens that cuda code is much slower to compile so the top of the list is all cuda files. Somewhat interestingly, `GridSample.cpp` was above `GridSample.cu` in compile time which is why that PR changes both.

The top of the list at the moment looks like this:
```
caffe2/CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/native/cuda/DistributionBernoulli.cu.o
     1m 13s 344ms
caffe2/CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/native/cuda/ForeachUnaryOp.cu.o
     1m 12s 425ms
caffe2/CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/native/cuda/Distributions.cu.o
     1m 8s 582ms
caffe2/CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/native/cuda/Indexing.cu.o
     1m 7s 595ms
caffe2/CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/native/cuda/group_norm_kernel.cu.o
     0m 59s 931ms
caffe2/CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/RegisterCUDA.cpp.o
     0m 55s 472ms
```

> FWIW, I like the idea of being stringent on inputs rather than letting a loophole like this let the user get away with an invalid input.

I wouldn't say that applies here since `MAX_DIMS` is an implementation limitation not an invalid input. If, for example, `matmul` allowed empty tensors to have a shape mismatch then I would agree. 

>Changing topic altogether, do you think that splitting code up this way causes any meaningful harm by creating cross-module optimization barriers?

I don't think there's much the compiler can do here, but things like calling the same tensor method in both functions will have some impact (especially for virtual methods). However, generally speaking, the heavy lifting of these functions are done in the cuda runtime to actually launch the kernel. So, if there is any slow down I expect it to be fairly minimal.

Commit:
Update on "Remove native_functions.yaml dependency from TensorTopK.cu"


Differential Revision: [D31856104](https://our.internmc.facebook.com/intern/diff/D31856104)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from TensorTopK.cu"


Differential Revision: [D31856104](https://our.internmc.facebook.com/intern/diff/D31856104)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856104).
